enhancements/installer/compact-plus-worker.md

```diff
@@ -0,0 +1,194 @@
+---
+title: Compact + Worker deployment as day0 operation
+authors:
+  - "@flaper87"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+api-approvers: # in case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers)
+  - TBD
+creation-date: 2022-03-08
+last-updated: 2022-03-08
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - TBD
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Compact + Worker deployment as day0 operation
+
+## Summary
+
+This enhancement proposes to allow for deploying a 4-nodes cluster as a day 0
+operation, where the infrastructure workload can be deployed with HA enabled.
+The 4-nodes cluster would be composed by 3 schedulable master nodes and a
+single worker, which provide users with a highly available cluster.
+
+## Motivation
+
+There are environments where only 4 nodes are available where 4-nodes clusters
+are being deployed. We could serve OpenShift cluster administrators of these
+environments better by allowing them to deploy a 4-nodes cluster as a day0
+operation and make sure their infrastructure workload is highly available.
+
+Today, it is possible to deploy a 4-nodes cluster. The resulting cluster will
+be functional but the infrastructure workload will be run as a SingleReplica
+topology, which won't allow the various services to run as HA.
+
+To achieve the desired topology described in this enhancement,OpenShift cluster
+administrators can only deploy a 4-nodes clusters by first deploying a compact
+cluster (3 schedulable masters) and then adding a host as a day 2 operation.
+Another option is for them to This provides the same results as aimed in this
+enhancement but with a poor and slower user experience.
+
+Today, because of the current limitations in the installer, it takes
+administrators significantly longer to achieve the desired topology, when only
+4 nodes are available. Furthermore, by requiring more steps, we are making the
+process more prone to user errors.
+
+### Goals
+
+List the specific goals of the proposal. How will we know that this has succeeded?
+
+- Enable the installer to deploy a 4-nodes cluster composed by 3 schedulable
+  masters and one worker.
+
+### Non-Goals
+
+- Define new infrastructure topologies
+
+- Define new architectures
+
+- Provide official support for the 4-nodes cluster deployments
```

Comment:
Then why are we doing this? Could the same not be achieved by having compute pool of size 1, modifying the manifests to make masters schedulable?

Reply:
I agree that this non-goal seems at direct odds with goal of the EP.

Reply:
I agree

Reply:
> For example, a user could achieve a 4 nodes cluster in 2 steps today. Does it make a 4 nodes cluster officially supported? Probably not.

I don't know why a 4-node cluster where the user added nodes after installation would not be supported. The user is certainly permitted to add and remove worker nodes post-installation.

Reply:
Perhaps it's worth considering test coverage - we have coverage of compact (3+0) and HA (3+2) clusters, do we have any coverage of the 3+1 case anywhere?  (I think the answer is no, but we may want to change that if we want to consider this configuration fully supported, or do we expect to only support/test in the context of ZTP?)

Reply:
> If we think this is specific to ZTP then it may be reasonable to just modify the installer-generated manifests (and this perhaps limits the test impact, e.g we probably don't want to test 3+1 for every platform in CI?)

I guess I don't where the line between supported and unsupported architectures is drawn.

FWIW, I am going to remove this line from the non-goals. I'm just taking it as an opportunity to think through what's really supported and what's not. 

I think @hardys  may be right in considering the architectures that we run tests on as the ones that are "officially" supported. 

As far as testing a 3+1 deployment like the one proposed here, I wonder if the 3+0 one that exists already would be enough. The difference between a 3+0 and a 3+1 would be that there is one node that is just a worker (instead of Master,Worker) so, in terms of architecture, they are both the same.

Reply:
IMO we should have at least some testing of the 3+1 topology, somewhere - testing 3+0 is not sufficient IMO since the worker hosts are deployed and potentially configured differently (different MCP/ignition, different deploy tooling in the IPI case) 

Also the runtime behavior is different to compact clusters - for example the ingress VIP can end up on either a controlplane or compute host, and may move between them

Reply:
We should remove this point from non-goals. If it is in non-goals there is very less motivation for implementing this.